### PR TITLE
raftstore: make batch strategy more fair

### PR DESCRIPTION
**Pull request by hicqu**
## What have you changed? (mandatory)
Make the batch strategy more fair.
Previously the batch system won't fetch more readies from other regions beside those currently in processing. So if there are many regions, and the write skews on just some regions, the other regions could be hungry.
This PR changes this behavior. Now the batch system can fetch messages from other regions, even if the current processing set is not finished.  

## What are the type of the changes? (mandatory)
Improvement.

## How has this PR been tested? (mandatory)
CI.

## Benchmark result if necessary (optional)
old result: 
default
new result:
default

### CODE REVIEW

File: src/raftstore/store/config.rs [hunk 1]
```diff
@@ -177,7 +177,7 @@ impl Default for Config {
             notify_capacity: 40960,
             snap_mgr_gc_tick_interval: ReadableDuration::minutes(1),
             snap_gc_timeout: ReadableDuration::hours(4),
-            messages_per_tick: 4096,
+            messages_per_tick: 128,
```
Comment (siddontang) [inline src/raftstore/store/config.rs]: may using 128 reduce the batch size and cause more IOs?

File: src/raftstore/store/fsm/batch.rs [hunk 2]
```diff
@@ -294,20 +296,29 @@ struct Poller<N: Fsm, C: Fsm, Handler> {
...
+                Err(_) => return,
+            }
+        } else {
+            true
         };
-        let mut tried = 1;
+
         while pushed {
-            if tried < max_size {
+            if curr_batch_len < max_size {
```
Comment (BusyJay) [inline src/raftstore/store/fsm/batch.rs]: Use `batch.len()` instead. It's wrong as L306 doesn't increase the value.

File: src/raftstore/store/fsm/metrics.rs [hunk 3]
```diff
@@ -22,6 +22,12 @@ lazy_static! {
         exponential_buckets(1.0, 2.0, 20).unwrap()
     )
     .unwrap();
+    pub static ref BATCH_FETCH_INTERVAL: Histogram = register_histogram!(
+        "tikv_raftstore_batch_fetch_interval",
```
Comment (BusyJay) [inline src/raftstore/store/fsm/metrics.rs]: What's the purpose of this metrics? It looks strange as it is unpredictable. A large value is not always worse/better than a small value.

**Comment by BusyJay** [comment]
How does it perform under regular workload?

**Comment by hicqu** [comment]
For the normal cases (not hot point), the benchmark result of master branch:
default
And the result of the patch:
default
The workload contains 50% select and 50% update.

File: src/raftstore/store/fsm/metrics.rs (hunk 3 above)
Comment (hicqu) [reply src/raftstore/store/fsm/metrics.rs]: The metrics helps to locate the problem. Although for most situations the metrics is unpredictable (`500us` is not worse/better than `5ms`), but `5ms` is very different from `50ms`. I think the metrics is still helpful for this case.
Or we can know the raftstore thread is blocked at other place or here if proposal wait time grows up?

Comment (BusyJay) [reply src/raftstore/store/fsm/metrics.rs]: If it's about to recording the interval, then it should be recorded before fetching. If it's recorded before fetching, what's the difference from `PEER_RAFT_PROCESS_DURATION`?

**Comment by BusyJay** [comment]
What about pure updates? Selection is usually handled by local read thread, so it should not be affected too much.

File: src/raftstore/store/config.rs (hunk 1 above)
Comment (hicqu) [reply src/raftstore/store/config.rs]: In one benchmark case, the write operations of raft engine grows from 220K to 250K. So let's keep the original value, and use a metrics to monitor the message batch size.

File: src/raftstore/store/fsm/metrics.rs (hunk 3 above)
Comment (hicqu) [reply src/raftstore/store/fsm/metrics.rs]: Em... yes they are almost same. Thank you!

File: src/raftstore/store/fsm/batch.rs [hunk 4]
```diff
@@ -294,20 +293,28 @@ struct Poller<N: Fsm, C: Fsm, Handler> {
 
 impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
     fn fetch_batch(&self, batch: &mut Batch<N, C>, max_size: usize) {
-        let mut pushed = match self.fsm_receiver.recv() {
-            Ok(fsm) => batch.push(fsm),
-            Err(_) => return,
+        let curr_batch_len = batch.len();
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/fsm/batch.rs]: Please add comments here for this changing.

Comment (hicqu) [reply src/raftstore/store/fsm/batch.rs]: Fixed.

**Comment by hicqu** [comment]
@BusyJay for pure update workload, the result shows this PR won't introduce any QPS dropping.
The left is for the patch, and the right is for master.
default

**Comment by hicqu** [comment]
PTAL @zhangjinpeng1987 @BusyJay

**Comment by zhangjinpeng1987** [comment]
I think we should add a simulating test to bench the batch system, maybe another PR.
BTW, is it possible to use unit tests to discover such hungry problems?

File: src/raftstore/store/fsm/batch.rs [hunk 5]
```diff
@@ -351,12 +360,9 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
...
                 }
             }
-            if batch.is_empty() {
-                batch_size = cmp::min(batch_size + 1, self.max_batch_size);
-                self.fetch_batch(&mut batch, batch_size);
-            } else {
-                batch_size = cmp::max(batch_size - 1, 1);
-            }
+            // Fetch batch after every round is finished. It's helpful to protect regions
+            // begin hungry if some regions are hot points.
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/fsm/batch.rs]: s/begin/become

**Comment by hicqu** [comment]
PTAL @BusyJay @zhangjinpeng1987 thank you!

**Comment by zhangjinpeng1987** [comment]
@BusyJay Any problem for this new strategy?

**Pull request merged**